File: distributed-process/benchmarks/erlang/latency.erl
-module(latency).

-export([server/0, client/2, start_server/0, start_client/2]).

forever(What) ->
  What(),
  forever(What).

replicate(0, _) ->
  ok;
replicate(N, What) ->
  What(),
  replicate(N - 1, What).

server() ->
  forever(fun () -> 
            receive 
              Them -> Them ! pong 
            end 
          end). 

client(N, Them) ->
  replicate(N, fun() ->
                 {pingServer, Them} ! self(),
                 receive 
                   pong -> ok
                 end
               end),
  io:format("Did ~B pings", [N]).
 
start_server() ->
  register(pingServer, spawn(latency, server, [])).

start_client(N, Them) ->
  client(N, Them).

File: distributed-process/benchmarks/Latency.hs
import System.Environment
import Control.Monad
import Control.Applicative
import Control.Distributed.Process
import Control.Distributed.Process.Node
import Network.Transport.TCP (createTransport, defaultTCPParameters)
import Data.Binary (encode, decode)
import qualified Data.ByteString.Lazy as BSL

pingServer :: Process ()
pingServer = forever $ do
  them <- expect
  send them ()

pingClient :: Int -> ProcessId -> Process ()
pingClient n them = do
  us <- getSelfPid
  replicateM_ n $ send them us >> (expect :: Process ())
  liftIO . putStrLn $ "Did " ++ show n ++ " pings"

initialProcess :: String -> Process ()
initialProcess "SERVER" = do
  us <- getSelfPid
  liftIO $ BSL.writeFile "pingServer.pid" (encode us)
  pingServer
initialProcess "CLIENT" = do
  n <- liftIO $ getLine
  them <- liftIO $ decode <$> BSL.readFile "pingServer.pid"
  pingClient (read n) them

main :: IO ()
main = do
  [role, host, port] <- getArgs
  Right transport <- createTransport host port defaultTCPParameters
  node <- newLocalNode transport initRemoteTable
  runProcess node $ initialProcess role
